### PR TITLE
(JP-3015) Undersampling_correction step

### PR DESCRIPTION
**Pull request by dmggh**
Resolves [JP-3015](https://jira.stsci.edu/browse/JP-3015)

Closes #

This PR creates a new step in the detector1 pipeline between the jump and ramp_fitting steps to deal with the issues seen for undersampled NIRISS images.

This PR supersedes  PR #7396,  so I will be deleting that earlier PR.  A unit test for datasets being skipped because they have too few groups still needs to be added.
 
**Checklist for maintainers**
- [X] added entry in `CHANGES.rst` within the relevant release section
- [X] updated or added relevant tests
- [X] updated relevant documentation
- [X] added relevant milestone
- [X] added relevant label(s)
- [x] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
- [ ] Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

### CODE REVIEW

**Comment by hbushouse** [comment]
Regression test run started at https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/582/

**Review by hbushouse — COMMENTED** [review]
Code looks OK. Will wait for additional unit test before approving.

**Review by hbushouse — APPROVED** [review]

File: jwst/undersampling_correction/undersampling_correction.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+    Returns
+    -------
+    gdq : int, 4D array
+        updated group dq array
+    """
+    n_ints = gdq.shape[0]
+    n_grps = gdq.shape[1]
+    n_rows = gdq.shape[2]
+    n_cols = gdq.shape[3]
```
Comment (nden) [inline jwst/undersampling_correction/undersampling_correction.py:70]: This is probably a trivial performance hit but is really not necessary to access `shape` 4 times as python allows unpacking of tuple values in one call
```
n_ints, n_grps, n_rows, n_cols = gdq.shape
```

File: jwst/undersampling_correction/undersampling_correction.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    num_pix = n_cols * n_rows
+
+    lowest_exc_1d = np.zeros(n_cols*n_rows) + n_grps
```
Comment (nden) [inline jwst/undersampling_correction/undersampling_correction.py:73]: `n_cols*n_rows` was already computed and assigned to `num_pix`.

**Comment by dmggh** [comment]
Jenkins regression test job :
https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/583/

**Pull request merged**